ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -354,28 +383,60 @@ chainSyncClient tracer cfg btime (ClockSkew maxSkew)
         { candidateChain       = candidateChain'
         , candidateChainState  = candidateChainState'
         }
-      return requestNext
-
-    requestNext :: Consensus (ClientPipelinedStIdle Z) blk tip m
-    requestNext = SendMsgRequestNext
-      handleNext
-      (return handleNext) -- when we have to wait
-
-    handleNext :: Consensus (ClientStNext Z) blk tip m
-    handleNext = ClientStNext
+      return $ requestNext mkPipelineDecision0 Zero ourHead theirHead
+
+    requestNext :: MkPipelineDecision
+                -> Nat n
+                -> WithOrigin SlotNo -- Our head
+                -> tip               -- Their head
+                -> Consensus (ClientPipelinedStIdle n) blk tip m
+    requestNext mkPipelineDecision n ourHead theirHead = do
+        case (n, runPipelineDecision mkPipelineDecision n ourBlockNo theirBlockNo) of
+          (_zero, (Request, mkPipelineDecision')) ->
+            SendMsgRequestNext
+              (handleNext mkPipelineDecision' Zero ourHead)
+              (return $ handleNext mkPipelineDecision' Zero ourHead) -- when we have to wait
+          (_, (Pipeline, mkPipelineDecision')) ->
+            SendMsgRequestNextPipelined
+              (requestNext mkPipelineDecision' (Succ n) ourHead theirHead)
+          (Succ n', (CollectOrPipeline, mkPipelineDecision')) ->
+            CollectResponse
+              (Just $ SendMsgRequestNextPipelined $
+                requestNext mkPipelineDecision' (Succ n) ourHead theirHead)
+              (handleNext mkPipelineDecision' n' ourHead)
+          (Succ n', (Collect, mkPipelineDecision')) ->
+            CollectResponse
+              Nothing
+              (handleNext mkPipelineDecision' n' ourHead)
+      where
+        -- TODO: We fake 'BlockNo' for now, because we don't have easy access
+        -- to 'SlotNo'. We need to think about that and address it.
```

Comment:
it might be good to add a github issue for this if you're not planning to resolve it here

Reply:
We have access to our `BlockNo`, see #964. But not to *their* `BlockNo`. We could instantiate the tip with `(Point (Header blk), BlockNo)`. What do you think @coot?

Reply:
I just didn't want to have to worry about that plumbing (to get our own `BlockNo`). If you think it's not too difficult, let's do it (instantiating the protocol as you say).

Reply:
Done

Reply:
if this is done the todo can be removed (maybe it's just github showing it here though)

Reply:
It's gone. GitHub mentions "Outdated" on the diff.